reconcile/acs_rbac.py

```diff
@@ -0,0 +1,558 @@
+    name: str
+    description: str
+    clusters: list[str]
+    namespaces: list[dict[str, str]]
+
+
+class Permission(OidcPermissionAcsV1):
+    def __hash__(self) -> int:
+        return hash(self.name)
+
+
+class AcsRole(BaseModel):
+    """
+    AcsRole is derived from oidc-permission-v1
+    Due to A-I role-1 ability to reference multiple oidc-permissions
+    the A-I role names cannot be used for the roles defined within ACS
+    """
+
+    name: str
+    description: str
+    assignments: list[AssignmentPair]
+    permission_set_name: str
+    access_scope: AcsAccessScope
+    system_default: Optional[bool]
+
+    @classmethod
+    def build(cls, permission: Permission, usernames: list[str]) -> Self:
+        assignments = [
+            AssignmentPair(
+                # https://github.com/app-sre/qontract-schemas/blob/main/schemas/access/user-1.yml#L16
+                key="org_username",
+                value=u,
+            )
+            for u in usernames
+        ]
+
+        is_unrestricted_scope = not permission.clusters and not permission.namespaces
+
+        return cls(
+            name=permission.name,
+            description=permission.description,
+            assignments=assignments,
+            permission_set_name=PERMISSION_SET_NAMES[permission.permission_set],
+            access_scope=AcsAccessScope(
+                # Due to api restriction, additional Unrestricted scopes
+                # cannot be made.
+                # Therefore, desired scopes that meet unrestricted condition
+                # are treated as the system default 'Unrestricted'
+                name=DEFAULT_ADMIN_SCOPE_NAME
+                if is_unrestricted_scope
+                else permission.name,
+                description=DEFAULT_ADMIN_SCOPE_DESC
+                if is_unrestricted_scope
+                else permission.description,
+                # second arg is returned even if first arg == False
+                clusters=[cluster.name for cluster in (permission.clusters or [])],
+                # mirroring format of 'rules.includedNamespaces' in /v1/simpleaccessscopes response
+                namespaces=[
+                    {
+                        "clusterName": n.cluster.name,
+                        "namespaceName": n.name,
+                    }
+                    for n in (permission.namespaces or [])
+                ],
+            ),
+            system_default=False,
+        )
+
+    def diff_role(self, b: Self) -> bool:
+        return (
+            self.permission_set_name != b.permission_set_name
+            or self.access_scope != b.access_scope
+            or self.description != b.description
+        )
+
+    def is_unrestricted_scope(self) -> bool:
+        # empty cluster and namespaces attributes signifies unrestricted scope
+        return not self.access_scope.clusters and not self.access_scope.namespaces
+
+
+class AcsRbacIntegrationParams(PydanticRunParams):
+    thread_pool_size: int
```

Comment:
Is this field used anywhere?

Reply:
Thanks for catching. Initial implementation of CRUD operations for rules utilized this but later found support for batch and dropped threading for the calls.

Reply:
[addressed here](https://github.com/app-sre/qontract-reconcile/pull/3899/commits/657c3818a33f5342ce10506c131f61ab3fdf7426)